src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace/RazorWorkspaceListenerBase.cs

```diff
@@ -0,0 +1,246 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace;
+
+public abstract class RazorWorkspaceListenerBase : IDisposable
+{
+    private static readonly TimeSpan s_debounceTime = TimeSpan.FromMilliseconds(500);
+    private readonly CancellationTokenSource _disposeTokenSource = new();
+
+    private readonly ILogger _logger;
+    private readonly AsyncBatchingWorkQueue<Work> _workQueue;
+
+    // Use an immutable dictionary for ImmutableInterlocked operations. The value isn't checked, just
+    // the existance of the key so work is only done for projects with dynamic files.
+    private ImmutableDictionary<ProjectId, bool> _projectsWithDynamicFile = ImmutableDictionary<ProjectId, bool>.Empty;
+
+    private Stream? _stream;
+    private Workspace? _workspace;
+
+    internal record Work(ProjectId ProjectId);
+    internal record UpdateWork(ProjectId ProjectId) : Work(ProjectId);
+    internal record RemovalWork(ProjectId ProjectId, string IntermediateOutputPath) : Work(ProjectId);
+
+    public RazorWorkspaceListenerBase(ILogger logger)
+    {
+        _logger = logger;
+        _workQueue = new(TimeSpan.FromMilliseconds(500), UpdateCurrentProjectsAsync, EqualityComparer<Work>.Default, _disposeTokenSource.Token);
+    }
+
+    void IDisposable.Dispose()
+    {
+        if (_workspace is not null)
+        {
+            _workspace.WorkspaceChanged -= Workspace_WorkspaceChanged;
+        }
+
+        if (_disposeTokenSource.IsCancellationRequested)
+        {
+            return;
+        }
+
+        _disposeTokenSource.Cancel();
+        _disposeTokenSource.Dispose();
+
+        _stream?.Dispose();
+        _stream = null;
+    }
+
+    public void NotifyDynamicFile(ProjectId projectId)
```

Comment:
Can this method be called from multiple threads?

Reply:
yes

Reply:
what all methods in this type can be called from multiple threads?

Reply:
`NotifyDynamicFile` should be the only one. `EnsureInitialized` is not thread safe and not intended to be. `Workspace_WorkspaceChanged` is driven by workspace change events so ordered accordingly. `UpdateCurrentProjectsAsync` is driven by the work queue

Reply:
EnsureInitialized is protected by a lock: https://github.com/dotnet/roslyn/blob/602687a61e2ec18a1a2b4e498779a4037809c7e3/src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/RazorWorkspaceListenerInitializer.cs#L52

NotifyDynamicFile calls come from Roslyn's IDynamicFileInfoProvider, which guarantees non-overlapping calls I believe.

Reply:
Ah perfect, I was trying to figure out if `IDynamicFileInfoProvider` had any guarantees by digging through code

Reply:
Should probably invoke @jasonmalinowski at this point, in case I'm writing checks (cheques? 😛) he can't cash

Reply:
What restrictions are there about how this can be called from multiple threads? For example, if calls to this can interleave with `Dispose` then this method can null ref. Hard to review parts of this change without understanding thiese types of interacitons.

Reply:
Mind if we talk offline for a bit? Might be quicker to share context and learn more about concerns

Reply:
> What che(ck|que) do you think you're writing here that might be a problem?

Just wanted to validate my statement that calls to `IDynamicFileInfoProvider.GetDynamicFileInfoAsync` won't overlap

Reply:
You mean this one?

https://github.com/dotnet/roslyn/blob/9951cebdc4f5581a0a06c1355f9f8c033ba0cb70/src/Workspaces/Core/Portable/Workspace/Host/SourceFiles/IDynamicFileInfoProvider.cs#L19-L26

We absolutely will call that on multiple threads.

Reply:
Glad I checked!

Reply:
AFAICT tell we should be okay still. This code hasn't changed in this PR, just moved. I can do another pass later to see if we can better clean this up, as well as better hook up the `SolutionChanged` event that wasn't handled correctly before